mypy/checker.py

```diff
@@ -3745,6 +3745,8 @@ def find_isinstance_check(self, node: Expression
                     elif (isinstance(vartype, Instance) and
                             vartype.type.fullname() == 'builtins.type'):
                         vartype = self.named_type('builtins.object')
+                    elif (isinstance(vartype, TypeVarType)):
+                        vartype = vartype.upper_bound
```

Comment:
I don't think this is correct, and I am surprised this works. I think this should be placed as a _separate_ `if` before the union branch.

Also this function is already quite big, so may you can refactor all the above logic for `issubclass()` in a helper function/method?